### PR TITLE
Agg auto spill refine

### PR DESCRIPTION
**Pull request by windtalker**
### What problem does this PR solve?

Issue Number: ref #7738

Problem Summary:

### What is changed and how it works?
For parallel aggregation, when an aggregation is marked to spill, spill all threads even if expected memories already released.
### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: dbms/src/Interpreters/AggSpillContext.cpp [hunk 1]
```diff
@@ -77,11 +77,33 @@ Int64 AggSpillContext::getTotalRevocableMemoryImpl()
 
 Int64 AggSpillContext::triggerSpillImpl(Int64 expected_released_memories)
 {
-    for (size_t i = 0; i < per_thread_revocable_memories.size() && expected_released_memories > 0; ++i)
+    size_t checked_thread = 0;
+    for (size_t i = 0; i < per_thread_revocable_memories.size(); ++i)
```
Comment (SeaRise) [inline dbms/src/Interpreters/AggSpillContext.cpp:81]:
```suggestion
    size_t checked_thread = 0;
    for (checked_thread  < per_thread_revocable_memories.size(); ++checked_thread)
```

File: dbms/src/Interpreters/AggSpillContext.cpp [hunk 2]
```diff
@@ -77,11 +77,33 @@ Int64 AggSpillContext::getTotalRevocableMemoryImpl()
...
+        if (per_thread_auto_spill_status[i].compare_exchange_strong(old_value, AutoSpillStatus::NEED_AUTO_SPILL))
+        {
+            LOG_DEBUG(log, "Mark thread {} to spill, expect to release {} bytes", i, per_thread_revocable_memories[i]);
+        }
         expected_released_memories = std::max(expected_released_memories - per_thread_revocable_memories[i], 0);
+        if (expected_released_memories == 0)
+        {
+            checked_thread = i;
+            break;
+        }
```
Comment (SeaRise) [inline dbms/src/Interpreters/AggSpillContext.cpp:93]:
```suggestion
            break;
```

File: dbms/src/Interpreters/AggSpillContext.cpp [hunk 3]
```diff
@@ -77,11 +77,33 @@ Int64 AggSpillContext::getTotalRevocableMemoryImpl()
...
+    }
+    for (size_t i = checked_thread + 1; i < per_thread_revocable_memories.size(); ++i)
+    {
+        /// unlike sort and hash join, the implementation of current agg spill does not support partial spill, that is to say,
+        /// once agg spill is triggered, all the data will be spilled in the end, so here to spill the data if memory usage is large enough
+        if (spill_config.max_cached_data_bytes_in_spiller > 0 && per_thread_revocable_memories[i] >= static_cast<Int64>(spill_config.max_cached_data_bytes_in_spiller))
```
Comment (SeaRise) [inline dbms/src/Interpreters/AggSpillContext.cpp:99]: move `spill_config.max_cached_data_bytes_in_spiller > 0` outside the foreach?
```
if (spill_config.max_cached_data_bytes_in_spiller > 0)
{
    for (xxx)
}
```

File: dbms/src/Interpreters/AggSpillContext.cpp (hunk 1 above)
Comment (windtalker) [reply dbms/src/Interpreters/AggSpillContext.cpp:81]: Done

**Review by yibin87 — APPROVED** [review]
LGTM

**Review by SeaRise — APPROVED** [review]
lgtm

**Pull request merged**